core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -231,7 +238,29 @@ abstract class AbstractFetcherThread(name: String,
     }
   }
 
-  def addPartitions(partitionAndOffsets: Map[TopicPartition, Long]) {
+  def markPartitionsForTruncation(partitionAndOffsets: Map[TopicPartition, Long]) {
+    partitionMapLock.lockInterruptibly()
+    try {
+      val newStates = partitionStates.partitionStates.asScala.map { state =>
+        val maybeNeedTruncation = partitionAndOffsets.get(state.topicPartition()) match {
+          case Some(offset) =>
+            // Remember the truncation offset for this partition provided by the caller
+            // This is only used by ReplicaAlterDirThread when leader epoch is undefined
+            partitionTruncationOffsets.put(state.topicPartition(), offset)
+            // Mark this partition for truncation using leader epoch
+            PartitionFetchState(state.value.fetchOffset, state.value.delay, truncatingLog = true)
```

Comment:
@junrao One reason to let a partition go through leader epoch is that, if the leader replica is truncated to offset 10, append data up to offset 20, and truncated again to offset 15, first truncation offset 10 will be overwritten by the second truncation offset 20. In reality the window for this happening is probably small and we expect the truncation to happen almost immediately on the future replica after the leader is truncated. But ideally we would probably want to stay on the safe side and use leader epoch to make sure this does not cause any problem.

Another reason to apply leader epoch is to mimic what we are currently doing with the follower replica. Currently whenever leader replica is truncated, it is guaranteed that there is leadership transfer, and the replica will be removed from the ReplicaFetcherManager and added back in the follower. Thus the follower will always apply leader epoch if leader is truncated. So it seems to make sense to always truncate the future replica using leader epoch whenever the current replica is truncated.

Does this make sense? Also, is there any correct or performance concern if we always use leader epoch to truncate the future replica?

Reply:
@lindong28 : Very good point on double truncation. One way to get around this is to maintain the smallest truncation offset if the partition is marked for truncation multiple times. My concern of depending on leader epoch is that it won't apply if the message format is old.

Reply:
@junrao Good point. I have update the patch to use the smallest truncation offset.

Even if we use the smallest truncation offset, I think it is still safer to use leader epoch to truncate the future replica if leader epoch is available. This is because truncation offset is only recorded in the memory which may be lost if broker restarts. Does this make sense?

Reply:
@junrao I have rebased patch onto trunk. I will go through this patch end-to-end after we agree on how to truncate the future replica.

Reply:
@lindong28 : Thinking a bit more. We can use leader epoch to handle truncation since if the leader epoch doesn't exist, we fall back to HW and the truncation point should always be no lower than that. So, this should be safe.

Reply:
@junrao Great. Thanks for taking time to think through this. Since it is safe to use leader epoch and there is not performance concern with doing it, to simplify the implementation, I am going to update the patch to use only leader epoch or high watermark to truncate the future replica without recording the truncation offset. Does this sound OK?